src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -342,15 +338,64 @@ private static ImmutableList<ItemData>.Builder ComputeItems(LazyItemList lazyIte
 
                 ImmutableHashSet<string> currentGlobsToIgnore = globsToIgnoreStack == null ? globsToIgnore : globsToIgnoreStack.Peek();
 
+                Dictionary<string, UpdateOperation> itemsWithNoWildcards = new Dictionary<string, UpdateOperation>(StringComparer.OrdinalIgnoreCase);
+                bool addedToBatch = false;
+
                 //  Walk back down the stack of item lists applying operations
                 while (itemListStack.Count > 0)
                 {
                     var currentList = itemListStack.Pop();
 
+                    if (currentList._memoizedOperation.Operation is UpdateOperation op)
+                    {
+                        bool addToBatch = true;
+                        int i;
+                        for (i = 0; i < op.Spec.Fragments.Count; i++)
+                        {
+                            ItemSpecFragment frag = op.Spec.Fragments[i];
+                            if (MSBuildConstants.CharactersForExpansion.Any(frag.TextFragment.Contains))
+                            {
+                                // Fragment contains wild cards, items, or properties. Cannot batch over it using a dictionary.
+                                addToBatch = false;
+                                break;
+                            }
+
+                            string fullPath = FileUtilities.GetFullPath(frag.TextFragment, frag.ProjectDirectory);
+                            if (itemsWithNoWildcards.ContainsKey(fullPath))
+                            {
+                                // Another update will already happen on this path. Make that happen before evaluating this one.
+                                addToBatch = false;
+                                break;
+                            }
+                            else
+                            {
+                                itemsWithNoWildcards.Add(fullPath, op);
+                            }
+                        }
+                        if (!addToBatch)
+                        {
+                            // Remove items added before realizing we couldn't skip the item list
```

Comment:
Just a slightly more descriptive idea.
```suggestion
                            // We found an item with a wildcard, remove all fragments from the current list so we don't process them yet.
```

Reply:
Emphasizing this one.